vtex/faststore
# feat: Trade Policies

## What's the purpose of this pull request?
This PR uses `VTEX Session`  to add full support for trade policies on FastStore

## How it works? 
We currently support a single trade policy by store on FastStore. You can configure your default trade policy at `store.config.js`. 
Supporting multiple sales channels means we are able to automatically detect and change the channel based on multiple parameters, like user and urls. 

To make this auto detection possible, I made the `Session` context provider behave like the `Cart` provider. This means that it now accepts a new function called `onValidateSession`. This functions calls the backend to validate the session. The backend, then, changes/validates the session and updates values like channel, person, language, etc.
On any change to the channel (because the user is logged in, for instance) the session is updated and the user starts browsing on this new session. 

To make all work, I:
- Changed `user` for `person` on `Session` Context. Now person can be accessed by `session.person`
- Removed the `person` query, since the person data is updated automatically on Session context
- Renamed `updateSession` mutation in favor of `validateSession`
- Forwards user's querystring to VTEX Session request, so session s updated accordingly.

## How to test it?
Test1: 
1. Change the account on store.config.js from `storeframework` to `storecomponents`
2. Run locally with `yarn develop`
3. Visit `/apparel---accessories`. Make sure the list of products appear
4. Visit `/apparel---accessories?sc=2`. Make sure the product list changes.

Test2:
1. Change the account on store.config.js from `storeframework` to `storecomponents`
2. Run locally with `yarn develop`
3. Visit the product `FASTSHOP TESTE HUB HML` 
4. Now, visit this path with `?sc=2`. Make sure the currency changed

### Starters Deploy Preview
- https://github.com/vtex-sites/gatsby.store/pull/68
- https://github.com/vtex-sites/nextjs.store/pull/68

### Future work
Requests for search api should wait for session validation. Currently these starters don't wait, causing possible flickering on the screen, since some products may appear and disappear (due to session being revalidated). Future work should include waiting for session validation before any requests.
